File: python/move_element_to_end/move_element_to_end.py
# O(n) time | O(1) space, where n is the length of the input array
def moveElementToEnd(array, toMove):
    # Initialize two pointers, one at the start and one at the end of the array
    idx = 0
    lastIdx = len(array) - 1

    # Iterate through the array until the two pointers meet
    while idx < lastIdx:
        # Move the lastIdx pointer to the left if the element at lastIdx is equal to toMove
        while idx < lastIdx and array[lastIdx] == toMove:
            lastIdx -= 1

        # If the current element at idx is equal to toMove, swap it with the element at lastIdx
        if array[idx] == toMove:
            array[idx], array[lastIdx] = array[lastIdx], array[idx]

        # Move the idx pointer to the right
        idx += 1

    # Return the mutated array with all of the instances of toMove moved to the end
    return array

File: java/Medium/move_element_to_end/MoveElementToEnd.java
package main.java.com.example.moveelementtoend;

import java.util.List;

public class MoveElementToEnd {


	// O(n) time | O(1) space, where n is the length of the input array
    public static List<Integer> moveElementToEnd(List<Integer> array, int toMove) {
        // Initialize two pointers: one starting from the beginning (idx) and one from the end (lastIdx) of the array
        int idx = 0;
        int lastIdx = array.size() - 1;

        // Iterate through the list until the two pointers meet
        while (idx < lastIdx) {
            // Move the lastIdx pointer to the left if the element at lastIdx equals toMove
            while (idx < lastIdx && array.get(lastIdx) == toMove) {
                lastIdx--;
            }

            // If the element at idx equals toMove, swap it with the element at lastIdx
            if (array.get(idx) == toMove) {
                swap(idx, lastIdx, array);
            }
            // Move the idx pointer to the right
            idx++;
        }
        // Return the mutated array with all of the instances of toMove moved to the end
        return array;
    }

    // Helper method to swap the elements at indices i and j in a list
    public static void swap(int i, int j, List<Integer> array) {
        int temp = array.get(j);
        array.set(j, array.get(i));
        array.set(i, temp);
    }

}